docs/design/coreclr/jit/DeabstractionAndConditionalEscapeAnalysis.md

```diff
@@ -0,0 +1,800 @@
+
+But, not so fast. The `MoveNext()` for `<NaiveWhere>d__50<T>` is also not inlined, despite being "hot," and that causes the original allocation to escape as well. Why? The primary reason is that the `MoveNext()` contains a `try/finally` and the JIT is currently unable to inline methods with EH. But the method is also quite complicated, so it's not clear if the JIT would inline it even if there was no EH, and it is not so easy to see what else might trip things up after all that, but it looks like the internal IEenumerable's `GetEnumerator()` is invoked conditionally and we'd at least need to be able to prove we knew that this always happens on the first "dominating" call to `MoveNext()`. (Footnote 17)
+
+In my initial examples the JIT also did not inline `get_Current` despite it being very simple: GDV claims it does not have a good class guess. It turns out it was just an artifact of my particular test case&mdash;no element passed the filter so `MoveNext` returned false on the first call, so `get_Current` was never called and so had no class profile. (Footnote 18)
+
+The upshot is that our inner loop still has two interface calls:
+```asm
+G_M8017_IG05:  ;; offset=0x00AF
+       mov      rcx, gword ptr [rbp-0x28]
+       mov      r11, 0x7FF8046202B0      ; code for System.Collections.IEnumerator:MoveNext():ubyte:this
+       call     [r11]System.Collections.IEnumerator:MoveNext():ubyte:this
+       test     eax, eax
+       je       SHORT G_M8017_IG07
+						;; size=21 bbWeight=1 PerfScore 5.50
+G_M8017_IG06:  ;; offset=0x00C4
+       mov      rcx, gword ptr [rbp-0x28]
+       mov      r11, 0x7FF8046202B8      ; code for System.Collections.Generic.IEnumerator`1[int]:get_Current():int:this
+       call     [r11]System.Collections.Generic.IEnumerator`1[int]:get_Current():int:this
+       add      esi, eax
+       jmp      SHORT G_M8017_IG05
+```
+(todo: take a closer look at `MoveNext`).
+
+### Linq Where
+
+Now let's briefly look at the `Linq` case ([link to source](https://github.com/dotnet/runtime/blob/4951e38fc5882ddf7df68c956acb6f5586ef47dd/src/libraries/System.Linq/src/System/Linq/Where.cs#L12)). Here things are similar but different. `Where` orchestrates handling for arrays and lists via special helper objects, and for other collection types relies on something like our `NaiveWhere`.
+
+For arrays the `ArrayWhereIterator` accesses the array directly rather than creating and using the standard array enumerator; this is presumably the reason for less allocation and faster iteration. The `MoveNext` does not contain EH and gets inlined, but again for the same reason as above `Current` is cold and so not inlined. (Footnote 18)
```

Comment:
```suggestion
For arrays the `ArrayWhereIterator` accesses the array directly rather than creating and using the standard array enumerator; this is presumably the reason for less allocation and faster iteration. The `MoveNext` does not contain EH and gets inlined, but again for the same reason as above `Current` is cold and so not inlined. (Footnote 19)
```